In [1]:
from datascience import *
import numpy as np
import math
%matplotlib inline

# Welcome to GWS-131's Data Science Module!

# Introduction to Python and Jupyter Notebooks:

## 1. Cells - Text and Code
In a notebook, each rectangle containing text or code is called a *cell*.

Cells (like this one) can be edited by double-clicking on them. This cell is a text cell, written in a simple format called [Markdown](http://daringfireball.net/projects/markdown/syntax) to add formatting and section headings.  You don't need to worry about Markdown today, but it's a pretty fun+easy tool to learn.

After you edit a cell, click the "run cell" button at the top that looks like ▶| to confirm any changes. (Try not to delete the instructions.) You can also press `SHIFT-ENTER` to run any cell or progress from one cell to the next.

Other cells contain code in the Python3 programming language.  Running a code cell will execute all of the code it contains.

Try running this cell:

In [2]:
print("Hello, World!")

Hello, World!


## 2. Numbers

Quantitative information arises everywhere in data science. In addition to representing commands to print out lines, expressions can represent numbers and methods of combining numbers. The expression `3.2500` evaluates to the number 3.25. (Run the cell and see.)

In [3]:
3.2500

3.25

And this one:

In [4]:
print(3)
4
5

3


5

Notice that we don't necessarily need to `print`. When you run a notebook cell, if the last line has a value, then Jupyter helpfully prints out that value for you. However, it won't print out prior lines automatically. If you wish to print out multiple lines, then the `print` function is helpful!

## 3. Arithmetic
Many basic arithmetic operations are built in to Python.  The Data 8 textbook section on [Expressions](http://www.inferentialthinking.com/chapters/03/1/expressions.html) describes all the arithmetic operators used in the course.  The common operator that differs from typical math notation is `**`, which raises one number to the power of the other. So, `2**3` stands for $2^3$ and evaluates to 8. 

The order of operations is what you learned in elementary school, and Python also has parentheses.  For example, compare the outputs of the cells below. Use parentheses for a happy new year!

In [5]:
1+(6*5-(6*3))**2*((2**3)/4*7)

2017.0

## 4. Names
In natural language, we have terminology that lets us quickly reference very complicated concepts.  We don't say, "That's a large mammal with brown fur and sharp teeth!"  Instead, we just say, "Bear!"

Similarly, an effective strategy for writing code is to define names for data as we compute it, like a lawyer would define terms for complex ideas at the start of a legal document.

In Python, we do this with *assignment statements*. An assignment statement has a name on the left side of an `=` sign and an expression to be evaluated on the right.

In [6]:
twenty = (3 * 11 + 5) / 2 - 9

When you run that cell, Python first evaluates the expression on the right: it computes the value of the expression `(3 * 11 + 5) / 2 - 9 `, which is the number 10.  Then it assigns that value the name `twenty`.  At that point, the code in the cell is done running.

After you run that cell, the value 10 is bound to the name `twenty`:

In [7]:
twenty

10.0

## 5. Functions

    
One important form of an expression is the call expression, which first names a function and then describes its arguments. The function returns some value, based on its arguments. Some important mathematical functions are

| Function | Description                                                   |
|----------|---------------------------------------------------------------|
| `abs`      | Returns the absolute value of its argument                    |
| `max`      | Returns the maximum of all its arguments                      |
| `min`      | Returns the minimum of all its arguments                      |
| `pow`      | Raises its first argument to the power of its second argument |
| `round`    | Round its argument to the nearest integer                     |

Here are two call expressions that both evaluate to 3

    abs(2 - 5)
    max(round(2.8), min(pow(2, 10), -1 * pow(2, 10)))

All these expressions are **compound expressions**, meaning that they are actually combinations of several smaller expressions.  `2 + 3` combines the expressions `2` and `3` by addition.  In this case, `2` and `3` are called **subexpressions** because they're expressions that are part of a larger expression.

A **statement** is a whole line of code.  Some statements are just expressions, like the examples above, that can be broken down into its subexpressions which get evaluated individually before evaluating the statement as a whole.


### 5.1. Calling functions

The most common way to combine or manipulate values in Python is by calling functions. Python comes with many built-in functions that perform common operations.

For example, the `abs` function takes a single number as its argument and returns the absolute value of that number.  The absolute value of a number is its distance from 0 on the number line, so `abs(5)` is 5 and `abs(-5)` is also 5.

In [8]:
abs(5)

5

In [9]:
abs(-5)

5

Functions can be called as above, putting the argument in parentheses at the end, or by using "dot notation", and calling the function after finding the arguments, as in the cell immediately below.

In [10]:
nums = make_array(1,2,5) # a list of numbers, will be explained in more detail soon
nums.max()

5

In [11]:
max(nums)

5

## 6. Strings

A `string` is a type of data, usually composed of alphabetical characters. A string is always enclosed in single or double quotations.

You can create variables that hold strings, and you can create strings to be any sequence of letters, numbers, and special characters that you want:

In [12]:
"I'm a string!"

"I'm a string!"

There's nothing stopping a string from being a number, but you can't do normal numerical operations on them.

In [13]:
sentence = "I'm a string"
sentence*4

"I'm a stringI'm a stringI'm a stringI'm a string"

There are, however, some convenient functions that allow you to convert strings to numbers, and numbers to strings: 

In [14]:
zero = '0'
int(zero)

0

In [15]:
str(zero)

'0'

**Strings as function arguments**

String values, like numbers, can be arguments to functions and can be returned by functions.  The function `len` takes a single string as its argument and returns the number of characters in the string: its **len**-gth.  

Note that it doesn't count *words*. 

**Challenge**

Use `len` to find out the number of characters in the string `welcome` below:  

In [16]:
welcome = "Welcome to this class!"

# your code below


## 7. Importing code


Most programming involves work that is very similar to work that has been done before.  Since writing code is time-consuming, it's good to rely on others' published code when you can.  Rather than copy-pasting, Python allows us to **import** other code, creating a **module** that contains all of the names created by that code.

Python includes many useful modules that are just an `import` away.  We'll look at the `math` module as a first example. The `math` module is extremely useful in computing mathematical expressions in Python. 

Suppose we want to very accurately compute the area of a circle with radius 5 meters.  For that, we need the constant $\pi$, which is roughly 3.14.  Conveniently, the `math` module has `pi` defined for us:

In [17]:
import math
radius = 5
area_of_circle = radius**2 * math.pi
area_of_circle

78.53981633974483

`pi` is defined inside `math`, and the way that we access names that are inside modules is by writing the module's name, then a dot, then the name of the thing we want:

    <module name>.<name>
    
In order to use a module at all, we must first write the statement `import <module name>`.  That statement creates a module object with things like `pi` in it and then assigns the name `math` to that module.  Above we have done that for `math`. If you wish to use a module in more than one cell, however, you only need to import it once — the first time you want to use it.

In [18]:
# Calculating factorials.
math.factorial(5)

120

In [19]:
# Calculating square roots.
math.sqrt(5)

2.23606797749979

## 8. Arrays

Up to now, we haven't done much that you couldn't do yourself by hand, without going through the trouble of learning Python.  Computers are most useful when you can use a small amount of code to *do the same action* to *many things at once*.


**Arrays** are how we put many values in one place so that we can operate on them as a group. For example, if `billions_of_numbers` is an array of numbers, the expression

```python
.10 * billions_of_numbers
```

gives a new array of numbers that's the result of multiplying each number in `billions_of_numbers` by .10 (10%).  Arrays are not limited to numbers; we can also put all the words in a book into an array of strings.

Concretely, an array is a **collection of values of the same type**, like a column in an Excel spreadsheet. 

### 8.1. Making arrays
You can type in the data that goes in an array yourself, but that's not typically how programs work. Normally, we create arrays by loading them from an external source, like a data file.

First, though, let's learn how to do it the hard way. Execute the following cell so that all the names from the `datascience` module are available to you.

In [20]:
from datascience import *

Now, to create an array, call the function `make_array`.  Each argument you pass to `make_array` will be in the array it returns.  Run this cell to see an example:

In [21]:
my_array = make_array(0.125, 4.75, -1.3)
my_array

array([ 0.125,  4.75 , -1.3  ])

Each value in an array (in the above case, the numbers 0.125, 4.75, and -1.3) is called an *element* of that array.

Arrays themselves are also values, just like numbers and strings.  That means you can assign them names or use them as arguments to functions.

In [22]:
print(my_array)

[ 0.125  4.75  -1.3  ]


In [23]:
my_array.item(0)

0.125

Notice that we wrote .item(0), not .item(1), to get the first element. This is a weird convention in programming and computer science: indices start at 0 instead of 1. So the first thing in an array is item 0, the second is item 1, and so on. It is also described as the number of elements that appear before that item. So 3 is the index of the 4th item.

## 9. Creating Tables

If we don't have a spreadsheet file and are starting with nothing, first we need to make arrays. In the case of a table, we'll consider an array as either a row or a column. Let's make two arrays below that will become our columns, one for some of our U.S. presidents and one for the year they were born.

In [24]:
president_names = make_array("Jefferson", "Garfield", "Eisenhower", "Obama")
president_birth = make_array(1743, 1831, 1890, 1961)

Now, to make a table using these arrays, we use the general form:

```python
Table( ).with_columns("Column Name", array_name, . . .)
```

We assign the created table to a variable (just like the arrays from above), and then type that variable name to display the table. 

In [25]:
pres_table= Table().with_columns("President", president_names,
                                  "Birth Year", president_birth)
pres_table 

President  | Birth Year
Jefferson  | 1743
Garfield   | 1831
Eisenhower | 1890
Obama      | 1961

Note that `with_columns` can also be used to add additional columns to existing tables, replacing `Table()` with the name of the table you want to add columns to.

In [26]:
pres_with_pets = pres_table.with_columns("Pet Name", make_array("Buzzy", "Veto", "Heidi", "Bo"))
pres_with_pets

President  | Birth Year | Pet Name
Jefferson  | 1743       | Buzzy
Garfield   | 1831       | Veto
Eisenhower | 1890       | Heidi
Obama      | 1961       | Bo

### 9.1 Importing

It's more likey that a file holding your data already exists. In general, to import data from a file, we write:

```python
Table.read_table("file_name")
```

Most often, these file names end in `.csv` to show the data format. `.csv` format is popular for spreadsheets and can be imported/exported from programs such as Microsoft Excel, OpenOffice Calc, or Google spreadsheets. 
 
An example is shown below using [U.S. Census data](http://www2.census.gov/programs-surveys/popest/datasets/2010-2015/national/asrh/nc-est2015-agesex-res.csv). 

In [27]:
census_data = Table.read_table("http://www2.census.gov/programs-surveys/popest/datasets/2010-2015/national/asrh/nc-est2015-agesex-res.csv")
census_data

SEX  | AGE  | CENSUS2010POP | ESTIMATESBASE2010 | POPESTIMATE2010 | POPESTIMATE2011 | POPESTIMATE2012 | POPESTIMATE2013 | POPESTIMATE2014 | POPESTIMATE2015
0    | 0    | 3944153       | 3944160           | 3951330         | 3963087         | 3926540         | 3931141         | 3949775         | 3978038
0    | 1    | 3978070       | 3978090           | 3957888         | 3966551         | 3977939         | 3942872         | 3949776         | 3968564
0    | 2    | 4096929       | 4096939           | 4090862         | 3971565         | 3980095         | 3992720         | 3959664         | 3966583
0    | 3    | 4119040       | 4119051           | 4111920         | 4102470         | 3983157         | 3992734         | 4007079         | 3974061
0    | 4    | 4063170       | 4063186           | 4077551         | 4122294         | 4112849         | 3994449         | 4005716         | 4020035
0    | 5    | 4056858       | 4056872           | 4064653         | 4087709         | 4132242         | 4123626         | 4006900         | 4018158
0    | 6    | 4066381       | 4066412           | 4073013         | 4074993         | 4097605         | 4142916         | 4135930         | 4019207
0    | 7    | 4030579       | 4030594           | 4043046         | 4083225         | 4084913         | 4108349         | 4155326         | 4148360
0    | 8    | 4046486       | 4046497           | 4025604         | 4053203         | 4093177         | 4095711         | 4120903         | 4167887
0    | 9    | 4148353       | 4148369           | 4125415         | 4035710         | 4063152         | 4104072         | 4108349         | 4133564
... (296 rows omitted)

That's a lot of information. As you can see from the labels on top, this table shows Biological Sex (0=total, 1=male, 2=female), Age,  2010 Census Information, and predictions for U.S. population for the next five years. 

### 9.2 Using Tables

We can make criteria to cut down tables. Accessing only the rows, columns, or values specfic to our purpose makes information easier understood. Analysis and conclusions can be made when data is more digestible. 

This notebook can calculate how large this table is with two functions: `num_rows` and `num_columns`. The general form for these functions are `<table>.num_rows` and `<table>.num_columns`. 

Let's use these on the table above. 

In [28]:
census_data.num_rows

306

In [29]:
census_data.num_columns

10

That's a 306 x 10 table! We can first start to cut down this table using only some columns. Let's only include biological sex, age and the estimated base for 2010 census data. 

   There are two methods to make a table with select columns included, `select` or `drop`:

- `select` can create a new table with only the columns indicated in the parameters 
- `drop` can create a new table with columns NOT indicated in the parameters


Here's an example of two equal lines: (keep in mind that we assign each new table to a new variable, to make organization easier). 

In [30]:
select_census_data = census_data.select("SEX", "AGE", "ESTIMATESBASE2010")
select_census_data

SEX  | AGE  | ESTIMATESBASE2010
0    | 0    | 3944160
0    | 1    | 3978090
0    | 2    | 4096939
0    | 3    | 4119051
0    | 4    | 4063186
0    | 5    | 4056872
0    | 6    | 4066412
0    | 7    | 4030594
0    | 8    | 4046497
0    | 9    | 4148369
... (296 rows omitted)

In [31]:
drop_census_data = census_data.drop("CENSUS2010POP","POPESTIMATE2010","POPESTIMATE2011","POPESTIMATE2012","POPESTIMATE2013","POPESTIMATE2014","POPESTIMATE2015")
drop_census_data

SEX  | AGE  | ESTIMATESBASE2010
0    | 0    | 3944160
0    | 1    | 3978090
0    | 2    | 4096939
0    | 3    | 4119051
0    | 4    | 4063186
0    | 5    | 4056872
0    | 6    | 4066412
0    | 7    | 4030594
0    | 8    | 4046497
0    | 9    | 4148369
... (296 rows omitted)

These two cells above give us the same resulting table, but `select` was easier to use since we only cared about a few columns. The resulting table was still pretty large (3x306), so our next step to cut it down even more is to only include non-gendered data AKA data where SEX=0, neither male or female specific.

To do this, we need to use a new function `where`. The general form of this function is:

```python
table.where(column_name, predicate)
```

To cut our table down to only include `sex=0`, we may use the predicate `are.equal_to()`. Note that we are assigning the new table to a new variable.

In [32]:
new_census_data = select_census_data.where("SEX", are.equal_to(0))
new_census_data

SEX  | AGE  | ESTIMATESBASE2010
0    | 0    | 3944160
0    | 1    | 3978090
0    | 2    | 4096939
0    | 3    | 4119051
0    | 4    | 4063186
0    | 5    | 4056872
0    | 6    | 4066412
0    | 7    | 4030594
0    | 8    | 4046497
0    | 9    | 4148369
... (92 rows omitted)

There are still 92 rows omitted! Let's take every 10th entry to cut this table down a little more. 

To do this we need to use the `take` function. The `take` function creates a new table with rows from the original table whose indices (row numbers) are given. Remember, in Python, indices start at 0! 

In [33]:
census_10_year = new_census_data.take([0,10,20,30,40,50,60,70,80,90])
census_10_year

SEX  | AGE  | ESTIMATESBASE2010
0    | 0    | 3944160
0    | 10   | 4172559
0    | 20   | 4519556
0    | 30   | 4285877
0    | 40   | 4383450
0    | 50   | 4660457
0    | 60   | 3621214
0    | 70   | 2043178
0    | 80   | 1308608
0    | 90   | 435695

Now that sex is all the same, we can drop that column. 

In [34]:
final_census_table = census_10_year.drop("SEX")
final_census_table

AGE  | ESTIMATESBASE2010
0    | 3944160
10   | 4172559
20   | 4519556
30   | 4285877
40   | 4383450
50   | 4660457
60   | 3621214
70   | 2043178
80   | 1308608
90   | 435695

---

#### Tables Essentials!

For your reference, here's a table of all the useful `Table` functions we've used so far:

|Name|Example|Purpose|
|-|-|-|
|`Table`|`Table()`|Create an empty table, usually to extend with data|
|`Table.read_table`|`Table.read_table("my_data.csv")`|Create a table from a data file|
|`with_columns`|`tbl = Table().with_columns("N", np.arange(5), "2*N", np.arange(0, 10, 2))`|Create a copy of a table with more columns|
|`column`|`tbl.column("N")`|Create an array containing the elements of a column|
|`sort`|`tbl.sort("N")`|Create a copy of a table sorted by the values in a column|
|`where`|`tbl.where("N", are.above(2))`|Create a copy of a table with only the rows that match some *predicate*|
|`num_rows`|`tbl.num_rows`|Compute the number of rows in a table|
|`num_columns`|`tbl.num_columns`|Compute the number of columns in a table|
|`select`|`tbl.select("N")`|Create a copy of a table with only some of the columns|
|`drop`|`tbl.drop("2*N")`|Create a copy of a table without some of the columns|
|`take`|`tbl.take(np.arange(0, 6, 2))`|Create a copy of the table with only the rows whose indices are in the given array|

### 9.3 Visualizations

Now that we have a manageable table we can start making visualizations! Below is a handy chart with all of the visualization functions we'll be working with and how you call them.

|Plotting type | | Call structure |
|-|-|-|
|Scatter | | `table.scatter("x column", "y column")` |
|Line | | `table.plot("x column", "y column")` |
|Bar | | `table.bar("x column", "y column")` |
|Horiz. Bar | | `table.barh("x column", "y column")` |
|Histogram | | `table.hist("x axis", bins(optional), unit(optional))` |